### PR TITLE
supported think mode

### PR DESCRIPTION
**Pull request by jiapingW**
## Motivation

We found that for inference models such as the qwq-32b and qwen3 series, the input IDs they construct are inconsistent with the expected ones.
Using qwen3 as an example:
```python
    messages = [
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Who are you?"},
        {"role": "assistant", "content": "I am a model."},
        {"role": "user", "content": "What is your name?"},
        {"role": "assistant", "content": "My name is QwQ."},
    ]
    res = tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=False,
    )
    # response will be:
    # <|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\nWho are you?<|im_end|>\n<|im_start|>assistant\nI am a model.<|im_end|>\n<|im_start|>user\nWhat is your name?<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\nMy name is QwQ.<|im_end|>\n

    res = tokenizer.apply_chat_template(
        messages[:-1],
        tokenize=False,
        add_generation_prompt=True,
    ) + message[-1]["content"] + "<|im_end|>\n"

    # response will be:
    # <|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\nWho are you?<|im_end|>\n<|im_start|>assistant\nI am a model.<|im_end|>\n<|im_start|>user\nWhat is your name?<|im_end|>\n<|im_start|>assistant\nMy name is QwQ.<|im_end|>\n
```
We can see it add a `<think>\n\n</think>\n\n` token at the end of the last response. It corresponds to the case where qwen3 adopts non-think mode. This is also the current implementation, which causes the Eagle model to train tokens that should not be involved in calculating the loss. As for the currently open source qwen3-eagle data such as https://huggingface.co/datasets/Tengyunw/qwen3_8b_eagle3, the data obtained uses the thinking mode. This will cause the training data and the inference data to be misaligned.

Therefore, we introduce an additional parameter for the inference model: **is-think-mode** to determine whether to use the prompt processing related to think mode. Because different models' **apply_chat_template** handles think mode differently, we choose to splice prompt to complete the following mode. Besides, we modified the template in **template.py** to adapt the corresponding think input to different models.

You can use the following code under the processing.py to debug and verify the correctness of the assistant message processed and the corresponding entire conversation.
```python 
if __name__ == "__main__":
    from transformers import AutoTokenizer
    tokenizer = AutoTokenizer.from_pretrained("Qwen3/Qwen3-30B-A3B", trust_remote_code=True)
    messages = [[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Who are you?"},
        {"role": "assistant", "content": "I am a model."},
        {"role": "user", "content": "What is your name?"},
        {"role": "assistant", "content": "My name is QwQ."}
    ]]
    chat_template = TEMPLATE_REGISTRY.get('qwen3')
    max_length = 2048
    res = preprocess_conversations(tokenizer, messages, chat_template, max_length,is_think_mode=False)
    print('\033[31mtext:\033[0m\n',tokenizer.decode(res["input_ids"][0][0]))
    print('\033[31massistant_message:\033[0m\n',tokenizer.decode(res['input_ids'][0][res['loss_mask'][0]==1]))
```

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by jiapingW** [comment]
Not making these changes will not affect model training, but it will cause the training results to deviate from expectations. For example, the \<think\> token has never been seen in the QwQ-32B training data.

File: specforge/data/preprocessing.py [hunk 1]
```diff
@@ -71,11 +72,16 @@ def _apply_loss_mask_from_chat_template(
...
+        f"{chat_template.end_of_turn_token}"
     )
+    if is_think_mode:
+        assistant_message_separator = (
+            f"{chat_template.end_of_turn_token}{chat_template.assistant_think_header}"
+        )
+    else:
+        assistant_message_separator = (
+            f"{chat_template.end_of_turn_token}{chat_template.assistant_header}"
+        )
```
Comment (shuaills) [inline specforge/data/preprocessing.py:85]: `    assistant_message_separator = (
        f"{chat_template.end_of_turn_token}{chat_template.assistant_header}"
        f"{chat_template.end_of_turn_token}"
    )`
    Any reason for removing f"{chat_template.end_of_turn_token}"?

File: specforge/data/preprocessing.py [hunk 2]
```diff
@@ -154,18 +161,20 @@ def preprocess_conversations(
             if source[0]["role"] != "user":
                 # if the first message is not from user, skip it
                 source = source[1:]
-
+            assert len(source) % 2 == 0, "conversation have question without answer"
```
Comment (shuaills) [inline specforge/data/preprocessing.py]: if len(source) % 2 != 0:
        continue  # raise ValueError("odd number of turns: question without answer")

File: specforge/data/preprocessing.py [hunk 3]
```diff
@@ -154,18 +161,20 @@ def preprocess_conversations(
...
             convroles = ["user", "assistant"]
-            for j, sentence in enumerate(source):
+            for j, sentence in enumerate(source[:-1]):
```
Comment (shuaills) [inline specforge/data/preprocessing.py]: Why is the last element skipped?

File: specforge/data/preprocessing.py [hunk 4]
```diff
@@ -154,18 +161,20 @@ def preprocess_conversations(
...
                 messages.append({"role": role, "content": sentence["content"]})
 
             conversation = tokenizer.apply_chat_template(
                 messages,
                 tokenize=False,
-                add_generation_prompt=False,
+                add_generation_prompt=True,
+                enable_thinking=is_think_mode,
             )
+            conversation += source[-1]["content"] + chat_template.end_of_turn_token
```
Comment (shuaills) [inline specforge/data/preprocessing.py]: The last role should be assistant.

File: specforge/data/preprocessing.py (hunk 1 above)
Comment (jiapingW) [reply specforge/data/preprocessing.py:85]: I adopted the suggested code from gemini code; I added back the chat_template.end_of_turn_token, restoring it to the original assistant_message_separator.

File: specforge/data/preprocessing.py (hunk 2 above)
Comment (jiapingW) [reply specforge/data/preprocessing.py]: Thanks, I think your modifications are more robust.

File: specforge/data/preprocessing.py (hunk 3 above)
Comment (jiapingW) [reply specforge/data/preprocessing.py]: I want the last round of dialogue to be formatted with the correct conversation template, rather than directly appending the entire conversation during construction. Otherwise, the same issue described in the motivation will occur, leading to constructed input IDs that are inconsistent with expectations. Therefore, I adopted the approach of concatenating only the last round of dialogue to ensure that the "think" format of the final round is correct.

File: specforge/data/preprocessing.py (hunk 4 above)
Comment (jiapingW) [reply specforge/data/preprocessing.py]: We hope that the final round of dialogue is not constructed directly using apply_chat_template, but rather by concatenating user and assistant messages. This is because the logic apply_chat_template uses to process a complete conversation is inconsistent with the output produced when apply_chat_template is applied directly to only the user message in the think model.

**Comment by jiapingW** [comment]
Anyone can help review it? I think it's important for train think mode such as Qwen3-8B.

**Comment by shuaills** [comment]
Can you resolve the conflicts and fix the lints? Thanks
Also some tests failed.

**Comment by jiapingW** [comment]
> Can you resolve the conflicts and fix the lints? Thanks Also some tests failed.

Wait some minutes, I will fix these.

**Comment by jiapingW** [comment]
I have resolved the conflicts and fixed the lints.

**Comment by shuaills** [comment]
Can you rebase it?

**Comment by jiapingW** [comment]
> Can you rebase it?

ok.

**Comment by jiapingW** [comment]
> > Can you rebase it?
> 
> I have handled it.

**Comment by shuaills** [comment]
Hi, I think this is ready to merge and can you check the lint and CI please?

**Comment by jiapingW** [comment]
> Hi, I think this is ready to merge and can you check the lint and CI please?

Thanks, I have handled the lint and  tests are queued.

**Comment by zyksir** [comment]
@jiapingW Hi, I have similar changes in #239 
I use ```python
conversation = self.tokenizer.apply_chat_template(
        messages, tokenize=False, add_generation_prompt=False, **kwargs
)``` so that the same code works both for Qwen and GPT. since for different thinking models, the kwargs for apply chat template is different. If we only use `is_think_mode`, we need to change the code again for GPT-OSS(and all other thinking models),
For #239 , we only need to change `data process` and add `is_think_mode: True` in the data.jsonl. I see that you are using that branch as well. Please let me you if that does not work for you.

**Comment by jiapingW** [comment]
> @jiapingW Hi, I have similar changes in #239 I use `python conversation = self.tokenizer.apply_chat_template( messages, tokenize=False, add_generation_prompt=False, **kwargs )` so that the same code works both for Qwen and GPT. since for different thinking models, the kwargs for apply chat template is different. If we only use `is_think_mode`, we need to change the code again for GPT-OSS(and all other thinking models), For #239 , we only need to change `data process` and add `is_think_mode: True` in the data.jsonl. I see that you are using that branch as well. Please let me you if that does not work for you.

I understand that your design supports more models in think mode. I used https://github.dev/sgl-project/SpecForge/tree/feature/add_sgl_online to train qwen2.5-7b. I've generated data for the qwen3 series, but haven't used your version for training yet. However, I believe your implementation may have the following issues:
1. Due to chat template misalignment, your training tokens may have some issues, such as training tokens that aren't originally involved in training, such as \<think>\n<\think>\n
2. As shown in the example above, qwen3's chat_template differs in its logic for handling multiple complete conversations and generating responses in think and non-think modes.

Simply adding parameters would likely cause issues, so my changes above separate the processing of the final round of conversation from the previous ones.

**Comment by jiapingW** [comment]
> @jiapingW Hi, I have similar changes in #239 I use `python conversation = self.tokenizer.apply_chat_template( messages, tokenize=False, add_generation_prompt=False, **kwargs )` so that the same code works both for Qwen and GPT. since for different thinking models, the kwargs for apply chat template is different. If we only use `is_think_mode`, we need to change the code again for GPT-OSS(and all other thinking models), For #239 , we only need to change `data process` and add `is_think_mode: True` in the data.jsonl. I see that you are using that branch as well. Please let me you if that does not work for you.

You can use the code below to test:
```python
if __name__ == "__main__":
    from transformers import AutoTokenizer
    tokenizer = AutoTokenizer.from_pretrained("Qwen3-30B-A3B", trust_remote_code=True)
    messages = [[
        {"role": "user", "content": "Who are you?"},
        {"role": "assistant", "content": "I am a model."},
        {"role": "user", "content": "What is your name?"},
        {"role": "assistant", "content": "My name is QwQ."},
        {"role": "user", "content": "What is 1+1"},
        {"role": "assistant", "content": "=2."}
    ]]
    # chat_template = TEMPLATE_REGISTRY.get('qwen')
    max_length = 2048

    conversation = tokenizer.apply_chat_template(
            messages,
            tokenize=False,
            add_generation_prompt=False,
            enable_thinking=True,
        )

    print(conversation)
```
You will get the response: 
```python
<|im_start|>user\nWho are you?<|im_end|>\n<|im_start|>assistant\nI am a model.<|im_end|>\n<|im_start|>user\nWhat is your name?<|im_end|>\n<|im_start|>assistant\nMy name is QwQ.<|im_end|>\n<|im_start|>user\nWhat is 1+1<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\n=2.<|im_end|>\n
```
I think the last conversation is not a thinking format which violates the expected answer.

**Comment by zyksir** [comment]
I agree with you. I will think about how to fix this unexcepted think token.

**Comment by zyksir** [comment]
For the thinking template, the training will be much more complex. Since in the training, we can only train the last conversation. For example, 
User: Question1
Qwen: <Think1><Answer1>;
User: Question2
Qwen: <Think2><Answer2>;
Actually, when the Qwen generate answer for the second question, it won't take <Think1> as history. Anyway I will close this PR for now and leave the unsolved problems as issues.

**Comment by jiapingW** [comment]
> For the thinking template, the training will be much more complex. Since in the training, we can only train the last conversation. For example, User: Question1 Qwen: ; User: Question2 Qwen: ; Actually, when the Qwen generate answer for the second question, it won't take as history. Anyway I will close this PR for now and leave the unsolved problems as issues.

Yes, the fundamental reason is that qwen's tokenizer doesn't support this type of full conversation processing. It requires the role of the last message to be user in order to process it properly.

In addition, I'm developing support for hidden state generation for offline DP training and experimenting with enabling sglang to use a partial world size launch for online DP training. I'd like to know where I can join discussions on framework development directions.

**Comment by zyksir** [comment]
@jiapingW Thank you! I will reopen a new PR to solve <think> problem for Qwen. Since this is a special problem for Qwen and the solution should not be in `GeneralParser`, I will add you as contributor since I will load some of the code from this PR.

**Comment by zyksir** [comment]
@jiapingW A really funny thing I find: For `Qwen/Qwen3-30B-A3B`, it will output another <think> token which we want to discard; while for `Qwen/Qwen3-235B-A22B-Instruct-2507`, it won't, which means the current code works fine.
```python
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-30B-A3B", trust_remote_code=True)
# tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-235B-A22B-Instruct-2507", trust_remote_code=True)
messages = [[
    {"role": "user", "content": "Who are you?"},
    {"role": "assistant", "content": "I am a model."},
    {"role": "user", "content": "What is your name?"},
    {"role": "assistant", "content": "My name is QwQ."},
    {"role": "user", "content": "What is 1+1"},
    {"role": "assistant", "content": "=2."}
]]
# chat_template = TEMPLATE_REGISTRY.get('qwen')
max_length = 2048

conversation = tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=False,
    )

print(conversation)
```

**Comment by jiapingW** [comment]
> @jiapingW A really funny thing I find: For `Qwen/Qwen3-30B-A3B`, it will output another token which we want to discard; while for `Qwen/Qwen3-235B-A22B-Instruct-2507`, it won't, which means the current code works fine.
> 
> ```python
> from transformers import AutoTokenizer
> tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-30B-A3B", trust_remote_code=True)
> # tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-235B-A22B-Instruct-2507", trust_remote_code=True)
> messages = [[
>     {"role": "user", "content": "Who are you?"},
>     {"role": "assistant", "content": "I am a model."},
>     {"role": "user", "content": "What is your name?"},
>     {"role": "assistant", "content": "My name is QwQ."},
>     {"role": "user", "content": "What is 1+1"},
>     {"role": "assistant", "content": "=2."}
> ]]
> # chat_template = TEMPLATE_REGISTRY.get('qwen')
> max_length = 2048
> 
> conversation = tokenizer.apply_chat_template(
>         messages,
>         tokenize=False,
>         add_generation_prompt=False,
>     )
> 
> print(conversation)
> ```

I remember Qwen3-235b-a22b-instruct-2507 was not a think model, you can see it here https://huggingface.co/Qwen/Qwen3-235B-A22B-Instruct-2507.  It will still generate the \<think>\n\n\</think> token, which is something we don’t want to generate. The result is 
``` python
['<|im_start|>user\nWho are you?<|im_end|>\n<|im_start|>assistant\nI am a model.<|im_end|>\n<|im_start|>user\nWhat is your name?<|im_end|>\n<|im_start|>assistant\nMy name is QwQ.<|im_end|>\n<|im_start|>user\nWhat is 1+1<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\n=2.<|im_end|>\n']
```
